lib/codegen/src/nan_canonicalization.rs

```diff
@@ -0,0 +1,110 @@
+use ir::{DataFlowGraph, Function, Inst, InstBuilder, InstructionData, Opcode, Value};
+use ir::condcodes::FloatCC;
+use ir::immediates::{Ieee32, Ieee64};
+use ir::types;
+use ir::types::Type;
+use timing;
+
+// Canonical 32-bit and 64-bit NaN values.
+static CANON_32BIT_NAN: u32 = 0b01111111100000000000000000000001;
+static CANON_64BIT_NAN: u64 = 0b0111111111110000000000000000000000000000000000000000000000000001;
+
+/// Performs the NaN-canonicalization pass by identifying floating-point
+/// arithmetic operations, and adding instructions to replace the result
+/// with a canonical NaN value if the result of the operation was NaN.
+pub fn do_nan_canonicalization(func: &mut Function) {
+    let _tt = timing::canonicalize_nans();
+    let mut pos = FuncCursor::new(func);
+    while let Some(_ebb) = pos.next_ebb() {
+        while let Some(inst) = pos.next_inst() {
+            if is_fp_arith(&mut pos, inst) {
+                add_nan_canon_instrs(&mut pos, inst);
+            }
+        }
+    }
+}
+
+/// Returns true/false based on whether the instruction is a floating-point
+/// arithmetic operation. This ignores operations like `fneg`, `fabs`, or
+/// `fcopysign` that only operate on the sign bit of a floating point value.
+fn is_fp_arith(pos: &mut FuncCursor, inst: Inst) -> bool {
+    // FIXUP: Should `ceil`, `floor`, `trunc`, `nearest`, and
+    // immediate constants be considered as well?
+    match pos.func.dfg[inst] {
+        InstructionData::Unary { opcode, .. } => opcode == Opcode::Sqrt,
+        InstructionData::Binary { opcode, .. } => {
+            opcode == Opcode::Fadd || opcode == Opcode::Fsub || opcode == Opcode::Fmul ||
+                opcode == Opcode::Fdiv || opcode == Opcode::Fmin ||
+                opcode == Opcode::Fmax
+        }
+        InstructionData::Ternary { opcode, .. } => opcode == Opcode::Fma,
+        _ => false,
+    }
+}
+
+/// Patch instructions that may result in a NaN result with operations to
+/// identify and replace NaN's with a single canonical NaN value.
+fn add_nan_canon_instrs(pos: &mut FuncCursor, inst: Inst) {
+    // FIXUP: Are these unwraps safe? An arithmetic operation should never be
+    // found at the end of an EBB, if I understand this correctly.
+
+    // Select the result of the instruction, move to the next instruction.
+    let inst_res: Value = pos.func.dfg.first_result(inst);
+    let next_inst: Inst = pos.next_inst().unwrap();
+
+    // Insert a comparison function, and a canonical NaN constant. Select
+    // the constant value, and move forward to the next instruction.
+    let is_nan: Value = pos.ins().fcmp(FloatCC::NotEqual, inst_res, inst_res);
+    insert_nan_const(pos, inst);
+    let canon_nan_instr: Inst = pos.prev_inst().unwrap();
+    let canon_nan_res: Value = pos.func.dfg.first_result(canon_nan_instr);
+    pos.goto_inst(next_inst);
+
+    // Insert a select instruction to canonicalize the NaN value.
+    pos.ins().select(is_nan, canon_nan_res, inst_res);
```

Comment:
Ok, so now let's talk about wiring in the new result value :-). The tricky thing is that we don't want to rewire *all* uses of the old value, because we need the fcmp and select to continue to use the old value.

As I mentioned earlier, there are a few ways to do it, but here's what I think is best: You can use `dfg`'s `replace_result` to replace the result value of the arithmetic instruction with a new `Value`. Then you can use `pos.ins()`'s `with_result` to when inserting the `select` using the original `Value`.